### PR TITLE
Add verify-pyproject-license hook

### PR DESCRIPTION
**Pull request by KyleFromNVIDIA**
Contributes to #29.

### CODE REVIEW

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:30]: So you're just looping until you find a random number that isn't anywhere in the document? In principle this could fail forever with a specifically constructed file, but I guess the edge case is so remote that it's not worth guarding against an infinite loop here.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:41]: I don't know how fast this function is, but you already compute it once above so you might as well cache `copied_document.as_string()` so that you can both evaluate its length and find the placeholder without needing to call it again.

Comment (vyasr) [reply src/rapids_pre_commit_hooks/pyproject_license.py:41]: Actually, better idea. Since you are guaranteed that your placeholder is a unique string that does not exist in the file, why not skip the traversal of the TOML hierarchy for replacement and instead just do (roughly, this skips some steps)
```
doc_str = document.as_string()
copied = copy(doc_str)  # At this point it's just a string, so copying will be faster
original_val = document.get(...)  # This is pseudocode, you'll have to do the recursive traversal you showed above to get the first arg
copied.replace(original_val, placeholder)
begin_val = copied.find(str(placeholder))
end_loc = begin_loc + len(original_val)
```

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:52]: I think this should probably be configurable. For most of our packages I would guess that a missing license would indicate an error that should cause the linting to fail.

**Review by vyasr — DISMISSED** [review]
Overall looks good.

I'm fine merging this with only pyproject license support since that is the more immediate need (where we get bit by wheeltamer), but we probably want to 1) add tooling to also validate our conda meta.yaml files, and 2) add logic for validating other fields. For 2, other fields we should standardize:
1. Author name (NVIDIA Corporation)
2. Classifiers: some packages will want more classifiers than others, but there are some base ones that we ought to be able to standardize. For instance, making sure the Python versions are consistent.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py:30]: To collide with every single number between `0` and `1048576`, the `pyproject.toml` document would have to be a minimum of about a megabyte of nothing but numbers. While it's theoretically possible, it's not something I would ever expect to encounter in a RAPIDS project.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py:41]: Is `copied` supposed to be a string or a parsed TOML document?

If it's a string, replacing `original_val` with `placeholder` does not guard against `original_val` appearing multiple times, hence the reason for the placeholder in the first place. Otherwise, we might as well just do a search for `original_val`.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py:52]: I agree, I'll add a warning if it doesn't exist.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (raydouglass) [inline src/rapids_pre_commit_hooks/pyproject_license.py:23]: Nearly everything in RAPIDS is Apache 2, but anything `ucx` related uses [`BSD-3-Clause`](https://github.com/rapidsai/ucxx/blob/branch-0.38/LICENSE).

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (raydouglass) [reply src/rapids_pre_commit_hooks/pyproject_license.py:30]: There probably should be some guard to break the loop. It's certainly unlikely to be a problem in regular usage, but a malicious user could push a change and cause CI to churn a lot. Though I think there is an absolute time limit on jobs.

Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py:30]: Good point about malicious users. I'll add some kind of limit.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py:23]: I've added a set of acceptable licenses, defaulting to `Apache 2.0` if it doesn't match any of them.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py]: I've added warnings and replacements to add the license if none exists.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [reply src/rapids_pre_commit_hooks/pyproject_license.py:45]: That's fair, I didn't consider the possibility of the license appearing twice but it's certainly possible. In that case you're right that you need to do the replacement on the TOML document before you convert it to a string, and I revert to my original, simpler, suggestion: you should be able to only call `copied_document.as_string()` once. Also I still don't think you shouldn't need these three `len`s since IIUC you're basically computing the length of the original value, which you can compute after the recursive traversal of nodes before you assign placeholder.

**Comment by bdice** [comment]
Just as a survey of various Python build tools:
- [Poetry says](https://python-poetry.org/docs/pyproject/#license) to use `license` with SPDX identifiers (so `Apache-2.0`, not `Apache 2.0`)
- [Python Packaging User Guide says](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license) don't use `license` if your project's license is well-known, use classifiers like `License :: OSI Approved :: Apache Software License` instead
  - setuptools offers similar guidance ([see this PR](https://github.com/pypa/setuptools/pull/1776))

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (bdice) [inline src/rapids_pre_commit_hooks/pyproject_license.py:33]: I'm super confused about what this is doing. Why are we generating random values to find a string?

Comment (bdice) [reply src/rapids_pre_commit_hooks/pyproject_license.py:33]: (I read [this previous thread](https://github.com/rapidsai/pre-commit-hooks/pull/30#discussion_r1586916941) but I still didn't understand the purpose of this.)

Comment (bdice) [reply src/rapids_pre_commit_hooks/pyproject_license.py:33]: If we're trying to insert a "garbage" string into the parsed representation in order to find the character index location for that key, we should avoid random integers and instead use a function like [uuid.uuid4()](https://docs.python.org/3/library/uuid.html#uuid.uuid4). We should do this as a one-shot and assume it will work, rather than retrying 10000 times.

Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py]: Yes, that's what we're trying to do. I've replaced the random integer with a UUID as you suggested.

**Comment by jakirkham** [comment]
It looks like all review requests were resolved. Requesting a new round of reviews

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:25]: Can we use the `RAPIDS_LICENSE` variable here to avoid duplicating?

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:32]: Can we add a comment indicating why we need this placeholder-based solution?

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:50]: There's some duplication here that we could clean up. What if instead of try-excepts, we check for the project key, and then the license key, and in each case set a flag `key_missing` or so and combine that with `project_table.is_super_table` to indicate that we insert the value from scratch?

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (vyasr) [inline src/rapids_pre_commit_hooks/pyproject_license.py:83]: Can this use `find_value_location`? Just discard the end afterwards since you're only using the beginning below.

**Review by vyasr — CHANGES_REQUESTED** [review]
Some clean up and deduplicating, otherwise looks good.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py]: I've modified `find_value_location()` to support the needed use case here.

**Review by vyasr — APPROVED** [review]

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (bdice) [inline src/rapids_pre_commit_hooks/pyproject_license.py:48]: Compute the placeholder once, above, rather than in both branches.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (bdice) [inline src/rapids_pre_commit_hooks/pyproject_license.py:89]: Do we want to replace/fix in this case? I think an unacceptable license should be a hard error. We don’t want to accidentally change licenses with pre-commit. License changes should be a deliberate action.

**Review by bdice — COMMENTED** [review]
A couple comments - please resolve them as you see fit. You can merge without waiting for more review from me if you wish.

File: src/rapids_pre_commit_hooks/pyproject_license.py
Comment (KyleFromNVIDIA) [reply src/rapids_pre_commit_hooks/pyproject_license.py]: Good point. I've taken out the replacement.

**Pull request merged**
